plugin/role_set.go

```diff
@@ -91,316 +101,256 @@ func (rs *RoleSet) bindingHash() string {
+
+	iamAdmin, err := b.IAMAdminClient(req.Storage)
+	if err != nil {
+		return []string{err.Error()}
 	}
-	return walId, nil
-}
 
-func (rs *RoleSet) updateIamPolicies(ctx context.Context, s logical.Storage, enabledIamResources iamutil.IamResourceParser, iamHandle *iamutil.IamHandle, rb ResourceBindings) ([]string, error) {
-	wals := make([]string, 0, len(rb))
-	for rName, roles := range rb {
-		walId, err := framework.PutWAL(ctx, s, walTypeIamPolicy, &walIamPolicy{
-			RoleSet: rs.Name,
-			AccountId: gcputil.ServiceAccountId{
-				Project:   rs.AccountId.Project,
-				EmailOrId: rs.AccountId.EmailOrId,
-			},
-			Resource: rName,
-			Roles:    roles.ToSlice(),
-		})
-		if err != nil {
-			return wals, err
-		}
+	iamHandle := iamutil.GetIamHandle(httpC, useragent.String())
 
-		resource, err := enabledIamResources.Parse(rName)
-		if err != nil {
-			return wals, err
+	warnings := make([]string, 0)
+	if boundResources.tokenGen != nil {
+		if err := b.deleteTokenGenKey(ctx, iamAdmin, boundResources.tokenGen); err != nil {
+			w := fmt.Sprintf("unable to delete key under service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.ResourceName(), err)
+			warnings = append(warnings, w)
 		}
+	}
 
-		p, err := iamHandle.GetIamPolicy(ctx, resource)
-		if err != nil {
-			return wals, err
+	if merr := b.removeBindings(ctx, iamHandle, boundResources.accountId.EmailOrId, boundResources.bindings); merr != nil {
+		for _, err := range merr.Errors {
+			w := fmt.Sprintf("unable to delete IAM policy bindings for service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.EmailOrId, err)
+			warnings = append(warnings, w)
 		}
+	}
 
-		changed, newP := p.AddBindings(&iamutil.PolicyDelta{
-			Roles: roles,
-			Email: rs.AccountId.EmailOrId,
-		})
-		if !changed || newP == nil {
-			continue
-		}
+	if err := b.deleteServiceAccount(ctx, iamAdmin, boundResources.accountId); err != nil {
+		w := fmt.Sprintf("unable to delete service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.ResourceName(), err)
+		warnings = append(warnings, w)
+	}
 
-		if _, err := iamHandle.SetIamPolicy(ctx, resource, newP); err != nil {
-			return wals, err
-		}
-		wals = append(wals, walId)
+	// If resources were deleted, we don't need the WAL rollbacks we created for these resources.
+	if len(warnings) == 0 {
+		b.tryDeleteWALs(ctx, req.Storage, walIds...)
 	}
-	return wals, nil
+
+	return nil
 }
 
-func roleSetServiceAccountName(rsName string) (name string) {
+// generateAccountNameForRoleSet returns a new random name for a Vault service account based off roleset name and time.
+// Note this is the name rather than the full email (i.e. string before @)
+//
+// As an example, for roleset "my-role" this returns `vaultmy-role-1234613`
+//
+func generateAccountNameForRoleSet(rsName string) (name string) {
 	// Sanitize role name
 	reg := regexp.MustCompile("[^a-zA-Z0-9-]+")
```

Comment:
This would be better as top level variable, compiled only once (and panicking immediately instead of in the request time path).